kerastuner/tuners/multi_execution_tuner.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2019 The Keras Tuner Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"Tuner that runs multiple executions per Trial."
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from ..engine import trial as trial_lib
+from ..engine import tuner as tuner_module
+from ..engine import oracle as oracle_module
+from ..engine import hyperparameters as hp_module
+from ..abstractions.tensorflow import TENSORFLOW_UTILS as tf_utils
+
+import collections
+import copy
+import json
+import numpy as np
+import random
+from tensorflow import keras
+
+
+class MultiExecutionTuner(tuner_module.Tuner):
+    """A Tuner class that averages multiple runs of the process.
+
+
+
+    Args:
+        oracle: Instance of Oracle class.
+        hypermodel: Instance of HyperModel class
+            (or callable that takes hyperparameters
+            and returns a Model instance).
+        executions_per_trial: Int. Number of executions
+            (training a model from scratch,
+            starting from a new initialization)
+            to run per trial (model configuration).
+            Model metrics may vary greatly depending
+            on random initialization, hence it is
+            often a good idea to run several executions
+            per trial in order to evaluate the performance
+            of a given set of hyperparameter values.
+        **kwargs: Keyword arguments relevant to all `Tuner` subclasses.
+            Please see the docstring for `Tuner`.
+    """
+
+    def __init__(self,
+                 oracle,
+                 hypermodel,
+                 executions_per_trial=1,
+                 **kwargs):
+        super(MultiExecutionTuner, self).__init__(
+            oracle, hypermodel, **kwargs)
+        if isinstance(oracle.objective, list):
+            raise ValueError(
+                'Multi-objective is not supported, found: {}'.format(
+                    oracle.objective))
+        self.executions_per_trial = executions_per_trial
+        # This is the `step` that will be reported to the Oracle at the end
+        # of the Trial. Since intermediate results are not used, this is set
+        # to 0.
+        self._reported_step = 0
+
+    def on_epoch_end(self, trial, model, epoch, logs=None):
+        # Intermediate results are not passed to the Oracle, and
+        # checkpointing is handled via a `ModelCheckpoint` callback.
```

Comment:
I don't understand the difference in checkpointing strategy. Why is checkpointing handled via calls to `save_weights` in the `Tuner` class and via a callback here?

Reply:
Here we can take advantage of the `ModelCheckpoint` class because we are assuming a single-objective. The `ModelCheckpoint` will only checkpoint the best checkpoint seen so far across all executions and epochs

In the `Tuner` class, we checkpoint every epoch because we want to make less assumptions about how the Oracle scores trials and sets the best_time (and thus chooses checkpoints). This allows for multi-objective scoring etc.

Reply:
In both classes, `get_best_models` uses the Trial.best_time to reload the checkpoint, but for `MultiExecutionTuner` the best_time is always set to `0` since we don't report intermediate results